src/backend/distributed/commands/vacuum.c

```diff
@@ -73,13 +79,68 @@ static CitusVacuumParams VacuumStmtParams(VacuumStmt *vacstmt);
 void
 PostprocessVacuumStmt(VacuumStmt *vacuumStmt, const char *vacuumCommand)
 {
-	int relationIndex = 0;
-	List *vacuumRelationList = ExtractVacuumTargetRels(vacuumStmt);
-	List *relationIdList = NIL;
 	CitusVacuumParams vacuumParams = VacuumStmtParams(vacuumStmt);
+
+	List *relationIdList = VacuumRelationIdList(vacuumStmt, vacuumParams);
+
+	if (ContainsOnlyPostgresTables(relationIdList))
+	{
+		return;
+	}
+
+	if (vacuumParams.options & VACOPT_VACUUM)
+	{
+		/*
+		 * We commit the current transaction here so that the global lock
+		 * taken from the shell table for VACUUM is released, which would block execution
+		 * of shard placements.
+		 */
+		CommitTransactionCommand();
+		StartTransactionCommand();
+	}
+
+	/*
+	 * Here we get the relation list again because we might have
+	 * closed the current transaction.
+	 */
+	relationIdList = VacuumRelationIdList(vacuumStmt, vacuumParams);
```

Comment:
we already did this

Reply:
Yeah, though if we commit then the memory context gets reset I guess so we lose the list (since we allocated it with palloc). I guess as a solution we can allocate the memory in a different memory context which doesn't get reset with committing.

Reply:
So there are two things to be done here:
- We can allocate the list in a different memory context so that we don't call this method twice
- We keep the current code.

The downside of the second one is that we do the same operation twice. But also this is not performance critical code block and it seems simpler.

If we choose the first approach, what memory context would be convenient to use here?

Reply:
Also in case it is useful:


```
#0  MemoryContextReset (context=0x559165d537d0) at mcxt.c:137
#1  0x00007f509827b86d in CoordinatedTransactionCallback (event=<optimized out>, arg=<optimized out>)
    at transaction/transaction_management.c:278
#2  0x00005591649bef33 in CallXactCallbacks (event=event@entry=XACT_EVENT_COMMIT) at xact.c:3521
#3  0x00005591649c3220 in CommitTransaction () at xact.c:2193
#4  0x00005591649c3fd0 in CommitTransactionCommand () at xact.c:2926
#5  0x00007f509820def2 in PostprocessVacuumStmt (vacuumStmt=vacuumStmt@entry=0x559165d67c38,
    vacuumCommand=vacuumCommand@entry=0x559165cd3298 "vacuum a;") at commands/vacuum.c:110
#6  0x00007f509820d0db in multi_ProcessUtility (pstmt=0x559165d679c8,
    queryString=0x559165cd3298 "vacuum a;", context=PROCESS_UTILITY_TOPLEVEL, params=0x0, queryEnv=0x0,
    dest=0x559165cd41e8, completionTag=0x7ffcca4e8ab0 "") at commands/utility_hook.c:604
#7  0x0000559164c8eb6e in ProcessUtility (pstmt=pstmt@entry=0x559165cd40f0,
    queryString=<optimized out>, context=<optimized out>, params=<optimized out>,
    queryEnv=<optimized out>, dest=dest@entry=0x559165cd41e8, completionTag=0x7ffcca4e8ab0 "")
    at utility.c:356
#8  0x0000559164c8ae65 in PortalRunUtility (portal=portal@entry=0x559165dab258,
    pstmt=pstmt@entry=0x559165cd40f0, isTopLevel=isTopLevel@entry=true,
    setHoldSnapshot=setHoldSnapshot@entry=false, dest=dest@entry=0x559165cd41e8,
    completionTag=completionTag@entry=0x7ffcca4e8ab0 "") at pquery.c:1175
#9  0x0000559164c8bacf in PortalRunMulti (portal=portal@entry=0x559165dab258,
    isTopLevel=isTopLevel@entry=true, setHoldSnapshot=setHoldSnapshot@entry=false,
    dest=dest@entry=0x559165cd41e8, altdest=altdest@entry=0x559165cd41e8,
    completionTag=completionTag@entry=0x7ffcca4e8ab0 "") at pquery.c:1321
#10 0x0000559164c8c920 in PortalRun (portal=portal@entry=0x559165dab258,
    count=count@entry=9223372036854775807, isTopLevel=isTopLevel@entry=true,
    run_once=run_once@entry=true, dest=dest@entry=0x559165cd41e8, altdest=altdest@entry=0x559165cd41e8,
    completionTag=0x7ffcca4e8ab0 "") at pquery.c:796
#11 0x0000559164c889c6 in exec_simple_query (query_string=query_string@entry=0x559165cd3298 "vacuum a;")
    at postgres.c:1215
#12 0x0000559164c8aa78 in PostgresMain (argc=<optimized out>, argv=argv@entry=0x559165d6f408,
    dbname=<optimized out>, username=<optimized out>) at postgres.c:4247
#13 0x0000559164bf9d15 in BackendRun (port=port@entry=0x559165cd2610) at postmaster.c:4448
#14 0x0000559164bfd1b0 in BackendStartup (port=port@entry=0x559165cd2610) at postmaster.c:4139
#15 0x0000559164bfd4cf in ServerLoop () at postmaster.c:1704
#16 0x0000559164bfe8f9 in PostmasterMain (argc=3, argv=<optimized out>) at postmaster.c:1377
#17 0x0000559164b5568f in main (argc=3, argv=0x559165ccb010) at main.c:228
```